### ISSUE TITLE
Battery percentage shows N/A when discharging

### ISSUE THREAD

**Issue by turbot151** [copycat-killer/lain#250]
I'm using lain with awesomewm 3.5.9-1
First time using this, so I try putting battery widget on the right of wibox. But the battery capacity shows N/A when `bat_now.status == "Discharging"` . Only when it was plugged in, the `bat_now.perc` displays the battery charge.
Here is my lines for battery widget
``` lua
batwidget = lain.widgets.bat ({
    timeout = 15,
    battery = "BAT0"
    ac = "AC"
    settings = function()
        widget:set_markup(" " .. bat_now.perc .. "% " .. bat_now.status)
    end
})
```
`ls /sys/class/power_supply/` returns "AC" and "BAT0"

**Comment by copycat-killer** [comment]
Fist, you're missing some commas, this is a correct snippet:
```lua
batwidget = lain.widgets.bat ({
    timeout = 15,
    battery = "BAT0",
    ac = "AC",
    settings = function()
        widget:set_markup(" " .. bat_now.perc .. "% " .. bat_now.status)
    end
})
```

Second, you're probably experiencing the same issue as #218 .

While discharging, `cat` me the files reported [here](https://github.com/copycat-killer/lain/issues/218#issuecomment-236541624) and also check [acpi](https://github.com/copycat-killer/lain/issues/218#issuecomment-236641842) output to see if your battery is correctly detected. Which laptop are you using?

In the meantime you can try installing upower and using [this widget instead](https://github.com/copycat-killer/lain/issues/218#issuecomment-237317990).

**Comment by turbot151** [comment]
Oh yeah, I forgot to type in some commas, my bad. The config run normally. 
I'm using HP Pavillion dv4t 1100-CTO Notebook
Here is the acpi output
`Battery 0: Discharging, 75%, discharging at zero rate - will never fully discharge.`
And the rest of `ls /sys/class/power_supply/` ( Some of the value may change as I was `cat` these files)
capacity
> 73 

charge_full (energy_full)
> 4752000

charge_now (energy_now)
> 3279000

No current_now
present
> 1

status
> Discharging

voltage_now
> 10800000

In the power folder:
async
> disable

autosuspend_delay_ms
> cat: /sys/class/power_supply/BAT0/power/autosuspend_delay_ms: Input/output error

control
> auto

runtime_active_kids, runtime_active_time
> 0

runtime_enabled
> disabled

runtime_status
> unsupported

runtime_suspended_time, runtime_usage
> 0

**Comment by copycat-killer** [comment]
Your battery doesn't provide a `current_now`, that's why the battery widget can't work properly. Even acpi says something strange like "*discharging at zero rate - will never fully discharge*", meaning that some  important infos are missing.

I don't know why this happens, nor I know how you could fix it. Probably distro-related?

As I told you, the only solution is to use something different, like a `upower`:

``` lua
mybattery = lain.widgets.abase({
    cmd = "upower -i /org/freedesktop/UPower/devices/battery_BAT | sed -n '/present/,/icon-name/p'",
    settings = function()
        bat_now = {}
        for k, v in string.gmatch(output, '([%a]+[%a|-]+):%s*([%a|%d]+[,|%a|%d]+)') do
            if     k == "present"       then bat_now.present      = v
            elseif k == "state"         then bat_now.state        = v
            elseif k == "warning-level" then bat_now.warninglevel = v
            elseif k == "energy"        then bat_now.energy       = string.gsub(v, ",", ".") -- Wh
            elseif k == "energy-full"   then bat_now.energyfull   = string.gsub(v, ",", ".") -- Wh
            elseif k == "energy-rate"   then bat_now.energyrate   = string.gsub(v, ",", ".") -- W
            elseif k == "voltage"       then bat_now.voltage      = string.gsub(v, ",", ".") -- V
            elseif k == "percentage"    then bat_now.percentage   = tonumber(v)              -- %
            elseif k == "capacity"      then bat_now.capacity     = string.gsub(v, ",", ".") -- %
            elseif k == "icon-name"     then bat_now.icon         = v
            end
        end

        -- customize here
        widget:set_text("Bat: " .. bat_now.percentage .. " " .. bat_now.state)
    end
})
```

Try it and let me know.

**Comment by turbot151** [comment]
I tried it and is working perfectly.
Thanks

**Comment by copycat-killer** [comment]
I added [a section in the wiki](https://github.com/copycat-killer/lain/wiki/bat#the-na-battery-percentage-when-discharging).

I'd like to tackle the `current_now` missing problem, but I really don't know where I could start from.

**Comment by turbot151** [comment]
I just found a small issue in UPower script. When the battery drop below 10% , `bat_now.percentage` return nil, causing error in the widget. I haven't found this before because I didn't let the battery drop below 10%.

**Comment by copycat-killer** [comment]
Try adjusting the command in `cmd`, and let me know if you come up with a solution. If even `upower` can't work properly with your battery, I'm afraid we can only assign default values, so that `bat_now` variables are `"N/A"` when not available.

Use [this new version](https://github.com/copycat-killer/lain/wiki/abase#upower).

**Comment by turbot151** [comment]
I found the culprit. Since the `string.gmatch` can't fetch the percentage when it drop to 1 digit, so I edited the pattern - replace the `+` with `-` at the end of it:
`string.gmatch(output, '([%a]+[%a|-]+):%s*([%a|%d]+[,|%a|%d]-)')`

**Comment by copycat-killer** [comment]
[Wiki updated](https://github.com/copycat-killer/lain/wiki/abase#upower), thanks.

**Comment by cyxou** [comment]
Hi, recently I've updated awesome vm to v4 and had to tweak my old rc.lua. This old battery widget worked fine and always showed the battery percentage:

```lua
netwidget = lain.widgets.bat({  
  settings = function()  
    bat_perc = bat_now.perc
    if bat_perc == "N/A" then bat_perc = "Plug" end
    widget:set_markup(markup(gray, " Bat ") .. bat_perc .. " ")
  end  
})  
```

Now, I've updated to the latest powerarrow-darker theme and battery widget in it always shows `N/A`. More to this, my old widget shows `N/A` too. I decided to use abase widget instead of it but it does not work either. Having debugged a little, I've spotted that the `output` variable which should return result from the shell cmd is always emty. I've tested the `base` widget and it worked fine to me, the `output` variable always return the result of the cmd run and all I did is replace this line: 
```lua
mybattery = lain.widgets.abase({
```
with this:
```lua
mybattery = lain.widgets.base({
```
I also set timeout to 60 second for updating the widget. I wonder, are there any side effects to use `base` widget instead of `abase` in this context?

**Comment by copycat-killer** [comment]
@cyxou Are you sure "BAT0" is still your battery id? If not, you have to reset `battery` in the `bat` widget.

Regarding the `abase` solution, here it works fine, maybe you have an issue with the shell: try setting

```lua
awful.util.shell = "/bin/sh" -- or whatever shell you use wich accept the option "-c"
``` 

before the widgets section.

In any case, don't worry: `bat` is synchronous too (and can't be asynchronized), so you'll not see any reduction in performance if using `upower` with `base`.

**Comment by cyxou** [comment]
Hi @copycat-killer , thanks for your quick response. I've triple checked the battery id and it is still `BAT0`. No problem here.

Funny, I've just commented out my `base` battery widget and uncommented the original battery widget code from your theme file and damn it worked. I've been struggling with it until 5 AM yesterday and now it just works!!! Did not modify anything! =)) Anyway, thanks again for your answer ))

**Comment by cyxou** [comment]
Hell no, the battery widget is broken again. I don't know why, but I've noticed that it again shows `N/A%`. Could it be because of the low charge left? I'll have some test tomorrow or one of these days and get back with results.

**Comment by copycat-killer** [comment]
At least we know that it works, sometimes. 😄 

Try monitoring your BAT files in different charge states and see if they are all present or someone disappears.

**Comment by cyxou** [comment]
Hi again!

Made some debugging and can conclude that on my system I can't get the contents of the `current_now` file. More to this the `power_now` file is absent. Here is the contens of all the files under my `BAT0` folder:   

```sh
▶ for f in /sys/class/power_supply/BAT0/* ; do echo $f: ; cat $f ; done
/sys/class/power_supply/BAT0/alarm:
0
/sys/class/power_supply/BAT0/capacity:
58
/sys/class/power_supply/BAT0/capacity_level:
Normal
/sys/class/power_supply/BAT0/charge_full:
4480000
/sys/class/power_supply/BAT0/charge_full_design:
4480000
/sys/class/power_supply/BAT0/charge_now:
2624000
/sys/class/power_supply/BAT0/current_now:
cat: /sys/class/power_supply/BAT0/current_now: No such device
/sys/class/power_supply/BAT0/cycle_count:
0
/sys/class/power_supply/BAT0/device:
cat: /sys/class/power_supply/BAT0/device: Is a directory
/sys/class/power_supply/BAT0/manufacturer:
Hewlett-Packard
/sys/class/power_supply/BAT0/model_name:
Primary
/sys/class/power_supply/BAT0/power:
cat: /sys/class/power_supply/BAT0/power: Is a directory
/sys/class/power_supply/BAT0/present:
1
/sys/class/power_supply/BAT0/serial_number:

/sys/class/power_supply/BAT0/status:
Discharging
/sys/class/power_supply/BAT0/subsystem:
cat: /sys/class/power_supply/BAT0/subsystem: Is a directory
/sys/class/power_supply/BAT0/technology:
Li-ion
/sys/class/power_supply/BAT0/type:
Battery
/sys/class/power_supply/BAT0/uevent:
POWER_SUPPLY_NAME=BAT0
POWER_SUPPLY_STATUS=Discharging
POWER_SUPPLY_PRESENT=1
POWER_SUPPLY_TECHNOLOGY=Li-ion
POWER_SUPPLY_CYCLE_COUNT=0
POWER_SUPPLY_VOLTAGE_MIN_DESIGN=11400000
POWER_SUPPLY_VOLTAGE_NOW=11496000
POWER_SUPPLY_CHARGE_FULL_DESIGN=4480000
POWER_SUPPLY_CHARGE_FULL=4480000
POWER_SUPPLY_CHARGE_NOW=2624000
POWER_SUPPLY_CAPACITY=58
POWER_SUPPLY_CAPACITY_LEVEL=Normal
POWER_SUPPLY_MODEL_NAME=Primary
POWER_SUPPLY_MANUFACTURER=Hewlett-Packard
POWER_SUPPLY_SERIAL_NUMBER=
/sys/class/power_supply/BAT0/voltage_min_design:
11400000
/sys/class/power_supply/BAT0/voltage_now:
11496000
```

So to make the bat widget work for me, I had to tweak it a little bit: 
```diff
diff --git a/widgets/bat.lua b/widgets/bat.lua
index ea749d3..ebf85e3 100644
--- a/widgets/bat.lua
+++ b/widgets/bat.lua
@@ -120,20 +120,29 @@ local function worker(args)
         bat_now.ac_status = tonumber(first_line(string.format("%s%s/online", pspath, ac))) or "N/A"
 
         if bat_now.status ~= "N/A" then
+            if bat_now.status ~= "Full" and sum_rate_power == 0 and bat_now.ac_status == 1 then
+								bat_now.perc  = math.floor(math.min(100, (sum_energy_now / sum_energy_full) * 100))
+								bat_now.time  = "00:00"
+								bat_now.watt  = 0
+
             -- update {perc,time,watt} iff battery not full and rate > 0
-            if bat_now.status ~= "Full" and (sum_rate_power > 0 or sum_rate_current > 0) then
+            elseif bat_now.status ~= "Full" then
                 local rate_time = 0
-                local div = (sum_rate_power > 0 and sum_rate_power) or sum_rate_current
 
-                if bat_now.status == "Charging" then
-                    rate_time = (sum_energy_full - sum_energy_now) / div
-                else -- Discharging
-                    rate_time = sum_energy_now / div
-                end
+                -- Calculate time and watt if rates are greater then 0
+                if (sum_rate_power > 0 or sum_rate_current > 0) then
+                    local div = (sum_rate_power > 0 and sum_rate_power) or sum_rate_current
 
-                if 0 < rate_time and rate_time < 0.01 then -- check for magnitude discrepancies (#199)
-                    rate_time_magnitude = math.abs(math.floor(math.log10(rate_time)))
-                    rate_time = rate_time * 10^(rate_time_magnitude - 2)
+                    if bat_now.status == "Charging" then
+                        rate_time = (sum_energy_full - sum_energy_now) / div
+                    else -- Discharging
+                        rate_time = sum_energy_now / div
+                    end
+
+                    if 0 < rate_time and rate_time < 0.01 then -- check for magnitude discrepancies (#199)
+                        rate_time_magnitude = math.abs(math.floor(math.log10(rate_time)))
+                        rate_time = rate_time * 10^(rate_time_magnitude - 2)
+                    end
                 end
 
                 local hours   = math.floor(rate_time)
                 local minutes = math.floor((rate_time - hours) * 60)
                 bat_now.perc  = math.floor(math.min(100, (sum_energy_now / sum_energy_full) * 100))
                 bat_now.time  = string.format("%02d:%02d", hours, minutes)
                 bat_now.watt  = tonumber(string.format("%.2f", sum_rate_energy / 1e6))
-            elseif bat_now.status ~= "Full" and sum_rate_power == 0 and bat_now.ac_status == 1 then
-								bat_now.perc  = math.floor(math.min(100, (sum_energy_now / sum_energy_full) * 100))
-								bat_now.time  = "00:00"
-								bat_now.watt  = 0
             elseif bat_now.status == "Full" then
                 bat_now.perc  = 100
                 bat_now.time  = "00:00"

``` 
The modified widget do not calculate time and watt values if it failed to obtain the data from either `current_now` or `power_now` files, but it now provides correct capacity percentage value. Could you check it out and test under different conditions? I can make a PR if you wish.

**Comment by copycat-killer** [comment]
@cyxou I tried you patch and implemented directly in d9877c4. Let me know if all is fine. Thanks.

**Comment by cyxou** [comment]
Man, thanks a lot! I've updated lain to the latest commit and everything seems to work fine for me now. 

I kindly appreciate for your lain library =)
